### PR TITLE
Re-order Test Summary Errors/Failures

### LINKED ISSUE #1044: Re-order Test Summary Errors/Failures
This is a little detail that has always bugged me but I have never fixed. Our current test summary looks like this,

```
Test Run Summary
    Overall result: Failed
   Tests run: 29, Passed: 28, Errors: 0, Failures: 1, Inconclusive: 0
     Not run: 0, Invalid: 0, Ignored: 0, Explicit: 0, Skipped: 0
  Start time: 2015-11-19 21:14:28Z
    End time: 2015-11-19 21:14:28Z
    Duration: 0.344 seconds
```

Our main purpose is Passed/Failed tests, but Errors come before Failures. I also think Failures should be renamed to Failed to match Passed.

I think it should look like this,

```
Test Run Summary
    Overall result: Failed
   Tests run: 29, Passed: 28, Failed: 1, Errors: 0, Inconclusive: 0
     Not run: 0, Invalid: 0, Ignored: 0, Explicit: 0, Skipped: 0
  Start time: 2015-11-19 21:14:28Z
    End time: 2015-11-19 21:14:28Z
    Duration: 0.344 seconds
```

If we do this NUnitLite will probably need to be changed too.

### LINKED ISSUE #1113: Console runner and xml output consistency
Currently I have a set of tests which are reporting the following in the console runner summary;

````
Test Run Summary
    Overall result: Failed
   Tests run: 59, Passed: 56, Errors: 1, Failures: 1, Inconclusive: 1
     Not run: 8, Invalid: 1, Ignored: 4, Explicit: 1, Skipped: 2
  Start time: 2015-12-03 04:27:39Z
    End time: 2015-12-03 04:27:39Z
    Duration: 0.489 seconds
````

The same test run produces the following test-run node in my XML output. 

````
<test-run id="2" testcasecount="67" result="Failed" total="67" passed="56" failed="3" inconclusive="1" skipped="7" asserts="11" engine-version="3.0.5797.27553" ..... >
````

After some thinking I can work out what the elements total, passed, failed, inconclusive, and skipped in the test-run node relate to in the console summary. The request is to simplify this process and place the same summary elements from the console into the XML node.

Thank you.

### PR DESCRIPTION
**Pull request by michaelhofer**
I have reworded and reordered the test results according to issue #1044 

Since this is my first contribution to the NUnit project, I hope that everything is correct. Please let me know if there is anything missing. As far as I can see, I have modified all necessary test cases.
I haven't changed the internal names yet, please let me know if I should do so as well or if you rather want to move this to a separate issue.

Fixes #1044 
Fixes #1113

### CODE REVIEW

**Comment by rprouse** [comment]
This looks good to me. The only issue I see is that @CharliePoole wanted to stick with the internal name Failure instead of Failed. I don't think that we can change the name internally now that we are released. Third party runners are likely relying on the values in the XML.

Should be an easy update to change the Failed back to Failure. If you push to your branch,l it will get added to this pull request. @CharliePoole, you agree that it should be switched back?

**Comment by michaelhofer** [comment]
@rprouse ok, thanks for the info. I suggest I will wait on @CharliePoole 's reponse. As soon as I get the ok to switch back, I will do this. This means, that my pull request then will only contain the modified order in the message, but no rewording (e.g. `Tests run: 29, Passed: 28, Failures: 1, Errors: 0, Inconclusive: 0`)

File: src/NUnitConsole/nunit3-console.tests/ResultReporterTests.cs [hunk 1]
```diff
@@ -101,7 +101,7 @@ public void SummaryReportTest()
             var expected = new string[] {
                 "Test Run Summary",
                 "    Overall result: Failed",
-                "   Tests run: 35, Passed: 30, Errors: 1, Failures: 3, Inconclusive: 1",
+                "   Tests run: 35, Passed: 30, Failed: 3, Errors: 1, Inconclusive: 1",
                 "     Not run: 10, Invalid: 3, Ignored: 4, Explicit: 3, Skipped: 0",
```
Comment (CharliePoole) [inline src/NUnitConsole/nunit3-console.tests/ResultReporterTests.cs]: Let's discuss what we want here. Using the internal terminology, there are seven tests with TestStatus.Failed:
 * 3 Failures "Failed" (ResultState.Failure)
 * 1 Error      "Failed:Error" (ResultState.Error)
 * 3 Invalid    "Failed:Invalid" (ResultState.NotRunnable)
That is, we currently have three different ways a test may be "Failed."

FWIW, we also have three ways a test may be skipped: "Skipped:Ignored", "Skipped:Explicit" and just plain "Skipped".

It's clear that our external description of this doesn't match the internal implementation perfectly, but maybe that's not a real problem. In a future update, maybe we should change this to first report the four basic status counts and then add other lines to break down those things that can be broken down. That is...

```
Tests run: 35, Passed: 30, Failed: 7, Inconclusive: 1, Skipped: 7
    Failed Tests - Failures: 3, Errors: 1, Invalid: 3
    Skipped Tests - Ignroed: 4, Explicit: 3, Other: 0
```

As I said, let's discuss before going further. Do we want to make our internals match user perception? Or do we want to push the users to understanding our internals?

**Comment by CharliePoole** [comment]
@michaelhofer Who knew that your first contribution would lead to such a deep discussion? :-)

We can go three ways on this - 1) change Failed back to Failures 2) Leave as is 3) Open up the entire can of worms.

Let's see what @rprouse says to my line comment.

**Comment by CharliePoole** [comment]
Something odd here. It's waiting for a status from the appveyor run on @michaelhofer 's branch. It may be that there is a conflict between his Appveyor settings and ours. Or maybe he has to sign up with Appveyor first. We should figure this out as it may imply some requirement on contributors as to how they set up in Appveyor. Alternatively, we could change our settings so that we only need a pr build to merge.

**Comment by CharliePoole** [comment]
@michaelhofer It's all good! You've called attention to two things that actually need attention!

**Comment by michaelhofer** [comment]
@CharliePoole yeah I didn't expect such a deep discussion :) I have signed up for Appveyor, but didn't add any projects there. Can you give my any hint how to proceed here? The regular Appveyor build as well as the Travis CI build completed. I don't know what to do about the third one, but have tried to add my repo in AppVeyor as well and triggered the checks again. Hopefully this will help.

**Comment by michaelhofer** [comment]
FYI: After adding my repo to AppVeyor, i was able to trigger a build. The tests passed and I have received a message shortly after the tests were done. However, there is no change here. Any ideas?

**Comment by rprouse** [comment]
Very odd that AppVeyor is not running or maybe not reporting the build for the branch. I thought it did that in the past. At any rate, the PR is equal to the branch. As far as I am concerned, people should not have to set up in AppVeyor to contribute to this project. The odd thing is, I contributed to Cake last week using a fork in the same way and my branch built. They are also using a protected branch.

As for the discussion. I don't feel that how we report the tests needs to match the internals, I think it just needs to be clear. The internal is just a implementation detail.

That said, I really like the proposed format,

```
Tests run: 35, Passed: 30, Failed: 7, Inconclusive: 1, Skipped: 7
    Failed Tests - Failures: 3, Errors: 1, Invalid: 3
    Skipped Tests - Ignored: 4, Explicit: 3, Other: 0
```

I think it makes it much clearer exactly what the results are. We might even skip the second or third line if there are no failed or skipped tests.

**Comment by michaelhofer** [comment]
@rprouse I think as well that the proposed format is much clearer. Shall I go on and change the output according to this and leave the internals as they are at the moment? I'm happy to help change that at a later time :)

**Comment by rprouse** [comment]
@michaelhofer don't worry about AppVeyor at your end. The fact that the PR builds and works is sufficient for me. If the protected master branch blocks us from merging, we will take care of it at our end.

As for changing to the proposed format, I like it. @CharliePoole what do you think?

**Comment by CharliePoole** [comment]
I'm in favor of my proposal. :-)

@michaelhofer I intended that the first line should reflect test status exactly. Since there are only four statuses, it has to add up to what we ran. You'll have to look at result state to see if it summarizes by status. The numbers should match exactly the corresponding attributes in the test-run element.

Don't be surprised if this takes another review or two. :-)

**Comment by CharliePoole** [comment]
@michaelhofer As a matter of logistics, the best way to handle this is to simply update your existing branch. The GitHub PR will update itself accordingly with the new commit and we will be able to see the aggregate changes under the Files Changed tab.

**Comment by michaelhofer** [comment]
I have now modified the structure and format of the reported results according to @CharliePoole 's proposed format. Please let me know if that suits your requirements. Are there any changes on the xml results necessary?

File: src/NUnitConsole/nunit3-console.tests/ResultReporterTests.cs [hunk 2]
```diff
@@ -101,8 +101,9 @@ public void SummaryReportTest()
...
                 "Test Run Summary",
                 "    Overall result: Failed",
-                "   Tests run: 35, Passed: 30, Errors: 1, Failures: 3, Inconclusive: 1",
-                "     Not run: 10, Invalid: 3, Ignored: 4, Explicit: 3, Skipped: 0",
+                "   Tests run: 35, Passed: 30, Failed: 7, Inconclusive: 1, Skipped: 7",
+                "     Failed Tests - Failures: 3, Errors: 1, Invalid: 3",
+                "     Skipped Tests - Ignored: 4, Explicit: 3, Other: 0",
                 "  Start time: 2015-10-19 02:12:28Z",
                 "    End time: 2015-10-19 02:12:29Z",
                 "    Duration: 0.349 seconds",
```
Comment (rprouse) [inline src/NUnitConsole/nunit3-console.tests/ResultReporterTests.cs]: It looks like we had it a bit messed up before, but it is more obvious now with the new format. I would like to see the indenting fixed here and in nunitlite. I would like Overall, Tests run and Start time indented 2 spaces and the nodes under them indented 4 spaces.

File: src/NUnitConsole/nunit3-console/ResultReporter.cs [hunk 3]
```diff
@@ -118,16 +118,21 @@ public void WriteSummaryReport()
 
             WriteSummaryCount("   Tests run: ", Summary.RunCount);
             WriteSummaryCount(", Passed: ", Summary.PassCount);
-            WriteSummaryCount(", Errors: ", Summary.ErrorCount, ColorStyle.Error);
-            WriteSummaryCount(", Failures: ", Summary.FailureCount, ColorStyle.Failure);
+            WriteSummaryCount(", Failed: ", Summary.FailedCount, ColorStyle.Failure);
             WriteSummaryCount(", Inconclusive: ", Summary.InconclusiveCount);
+            WriteSummaryCount(", Skipped: ", Summary.SkipCount);
```
Comment (rprouse) [inline src/NUnitConsole/nunit3-console/ResultReporter.cs]: Doesn't this need to be a new property, maybe TotalSkipCount which would be ignore + explicit + Skip? I am also not sure how the test suite above has 7 outputed here and 0 for the Other when they use the same variable?

File: src/NUnitConsole/nunit3-console.tests/ResultReporterTests.cs (hunk 2 above)
Comment (michaelhofer) [inline src/NUnitConsole/nunit3-console.tests/ResultReporterTests.cs]: @rprouse Sounds good to me. Sorry about the multiple commits, was struggling with some git issues earlier. I can squash them as soon as we have sorted all things out.
I will modify my changes according to your comment now.

File: src/NUnitConsole/nunit3-console/ResultReporter.cs (hunk 3 above)
Comment (michaelhofer) [inline src/NUnitConsole/nunit3-console/ResultReporter.cs]: @rprouse You are absolutely right. I have implemented the new property in my last commit. The tests are running at the moment, so please hang on a couple of minutes in order to see the latest status.

**Comment by michaelhofer** [comment]
@rprouse I have now pushed my latest changes to my branch and the tests passed. Could you please have another look? Just let me know if I have missed anything :)

**Comment by rprouse** [comment]
Looks good to me. @CharliePoole can you take a second look?

I am going to look at the master branch settings to see if we can merge with just our branches building on AppVeyor.

**Comment by rprouse** [comment]
@CharliePoole I updated the master branch protection settings. It is still a requirement that the PR build on AppVeyor and Travis. I removed the requirement that the branch build too. It is now mergable.

File: src/NUnitConsole/nunit3-console.tests/ResultReporterTests.cs [hunk 4]
```diff
@@ -100,9 +100,10 @@ public void SummaryReportTest()
         {
             var expected = new string[] {
                 "Test Run Summary",
-                "    Overall result: Failed",
-                "   Tests run: 35, Passed: 30, Errors: 1, Failures: 3, Inconclusive: 1",
-                "     Not run: 10, Invalid: 3, Ignored: 4, Explicit: 3, Skipped: 0",
+                "  Overall result: Failed",
+                "  Tests run: 35, Passed: 30, Failed: 7, Inconclusive: 1, Skipped: 7",
```
Comment (CharliePoole) [inline src/NUnitConsole/nunit3-console.tests/ResultReporterTests.cs]: The number 35 no longer makes sense here. Previously it was the sum of the other items on the line, that is passed + errors + failures + inconclusive. I guess it should now be the count of  test cases, which is 45, the total of the earlier tests run and not run. @rprouse What do you think?

File: src/NUnitConsole/nunit3-console/ResultSummary.cs [hunk 5]
```diff
@@ -74,6 +74,22 @@ public int NotRunCount
         }
 
         /// <summary>
+        /// Returns the number of failed test cases (including errors and invalid tests)
+        /// </summary>
+        public int FailedCount
+        {
+            get { return FailureCount + InvalidCount + ErrorCount;  }
```
Comment (CharliePoole) [inline src/NUnitConsole/nunit3-console/ResultSummary.cs]: At some (maybe future) point, I think we want to just add up the test cases with a status of failed. That way, when we add new kinds of failed tests, the total will continue to be correct. Same for the next one.

File: src/NUnitFramework/nunitlite.runner/TextUI.cs [hunk 6]
```diff
@@ -342,20 +342,25 @@ public void DisplaySummaryReport(ResultSummary summary)
                 SkipLine();
 
             WriteSectionHeader("Test Run Summary");
-            WriteLabelLine("   Overall result: ", overallResult, overallStyle);
+            WriteLabelLine("  Overall result: ", overallResult, overallStyle);
 
-            WriteSummaryCount("   Tests run: ", summary.RunCount);
+            WriteSummaryCount("  Tests run: ", summary.RunCount);
```
Comment (CharliePoole) [inline src/NUnitFramework/nunitlite.runner/TextUI.cs]: Let's make this Test Count: or Test Cases: and use the total of test cases as the value.

File: src/NUnitFramework/nunitlite.runner/TextUI.cs [hunk 7]
```diff
@@ -342,20 +342,25 @@ public void DisplaySummaryReport(ResultSummary summary)
...
+            WriteSummaryCount(", Failed: ", summary.FailedCount);
             WriteSummaryCount(", Inconclusive: ", summary.InconclusiveCount);
+            WriteSummaryCount(", Skipped: ", summary.SkipCount);
+            _outWriter.WriteLine();
+
+            _outWriter.Write("    Failed Tests - "); 
+            WriteSummaryCount("Failures: ", summary.FailureCount);
+            WriteSummaryCount(", Errors: ", summary.ErrorCount);
+            WriteSummaryCount(", Invalid: ", summary.InvalidCount);
             _outWriter.WriteLine();
```
Comment (CharliePoole) [inline src/NUnitFramework/nunitlite.runner/TextUI.cs]: This group of lines could be conditional on having some Failed tests.
If (summary.FailedCount > 0)...

File: src/NUnitFramework/nunitlite.runner/TextUI.cs [hunk 8]
```diff
@@ -342,20 +342,25 @@ public void DisplaySummaryReport(ResultSummary summary)
...
 
-            WriteSummaryCount("     Not run: ", summary.NotRunCount);
-            WriteSummaryCount(", Invalid: ", summary.InvalidCount, ColorStyle.Error);
-            WriteSummaryCount(", Ignored: ", summary.IgnoreCount, ColorStyle.Warning);
+            _outWriter.Write("    Skipped Tests - "); 
+            WriteSummaryCount("Ignored: ", summary.IgnoreCount);
             WriteSummaryCount(", Explicit: ", summary.ExplicitCount);
-            WriteSummaryCount(", Skipped: ", summary.SkipCount);
+            WriteSummaryCount(", Other: ", summary.SkipCount);
```
Comment (CharliePoole) [inline src/NUnitFramework/nunitlite.runner/TextUI.cs]: This group should depend on there actually being some skipped tests.

**Comment by CharliePoole** [comment]
I suggested further changes inline...

1. Replace Tests Run in summary by either Test Count or Test Cases with the total number of tests handled. 

2. Make the failed tests and skipped tests lines conditional on having something to display.

Otherwise it looks good. Nice work!

**Comment by michaelhofer** [comment]
@CharliePoole thanks for your comments, I have now updated my changes according your comments. The checks have passed, just let me know if there are some more comments from your side.

**Comment by rprouse** [comment]
Thanks for the contribution @michaelhofer. If you are interested in helping out more, we have marked a few other issues with the label `EasyFix`. They are a great way to get your feet wet.

@CharliePoole  and I will try to add the label to more issues as we plan for the 3.2 release over the next week.

**Pull request merged**